### PR TITLE
[Core] Cleaning up and adding templates to some methods on GeometryUtilities

### PR DESCRIPTION
**Pull request by loumalouomega**
This cleans the GeometryUtilities and adds template arguments to some functions (so bounded matrices can be used)

### CODE REVIEW

**Comment by loumalouomega** [comment]
Ping

**Comment by pablobecker** [comment]
Lots of changes here (high possiblity to add bugs?) 
do these functions have tests?

**Comment by loumalouomega** [comment]
Many of them not, I will add it

In any case the changes are adapting to style guide and adding missing const (and deprecating duplicated methods)

**Comment by loumalouomega** [comment]
Still some tests missing, on it. I fixed some things that weren't working even in the master

**Comment by loumalouomega** [comment]
@pablobecker Everything tested

**Comment by loumalouomega** [comment]
@roigcarlo can you check why this test fails in Travis?, in my machine both FullDebug and Release works fine

File: kratos/tests/cpp_tests/utilities/test_geometry_utils.cpp [hunk 1]
```diff
@@ -45,11 +121,215 @@ namespace Testing {
...
+        // Computing the info
+        array_1d<double, 4> distances;
+        distances[0] = 0.1;
+        distances[1] = 0.2;
+        distances[2] = -0.3;
+        distances[2] = 0.2;
+        array_1d<Point, 4> intersection_points;
+        const int intersections = GeometryUtils::CalculateTetrahedraIntersectionPoints(tetrahedra, distances, intersection_points);
+
+        KRATOS_CHECK_EQUAL(intersections, 1);
```
Comment (loumalouomega) [inline kratos/tests/cpp_tests/utilities/test_geometry_utils.cpp]: @roigcarlo Is this test
```suggestion
//        KRATOS_CHECK_EQUAL(intersections, 1);
```

**Comment by pooyan-dadvand** [comment]
I don't know how we arrive to this point having so many copy of the determinant and Jacobians over all Kratos... @RiccardoRossi this is our fault

Meanwhile too many changes in a basic tool

**Comment by loumalouomega** [comment]
> I don't know how we arrive to this point having so many copy of the determinant and Jacobians over all Kratos... @RiccardoRossi this is our fault
> 
> Meanwhile too many changes in a basic tool

I added tests to everything (in fact we didn't have tests previously)

File: kratos/utilities/geometry_utilities.h [hunk 2]
```diff
@@ -821,13 +676,17 @@ class GeometryUtils
...
-    static void DeformationGradient(Matrix const& rJ, Matrix const& rInvJ0, Matrix& rF)
+    template<class TMatrix1, class TMatrix2, class TMatrix3>
+    static void DeformationGradient(
+        TMatrix1 const& rJ,
+        TMatrix2 const& rInvJ0,
+        TMatrix3& rF
+        )
     {
         if (rF.size1() != rJ.size1() || rF.size2() != rInvJ0.size2())
             rF.resize(rJ.size1(), rInvJ0.size2(), false);
```
Comment (philbucher) [inline kratos/utilities/geometry_utilities.h:699]: this will not compile with AMatrix, same issue that was already discussed in the previous days

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:699]: True

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:699]: I "fixed" on https://github.com/KratosMultiphysics/Kratos/pull/4144/commits/6d86cb2446f895d464fd300ca5a4c8b467389219 sincerely is difficult to tackle that

**Comment by pablobecker** [comment]
@KratosMultiphysics/implementation-committee has reached consensus. These changes are correct and we are waiting for the final decision from the @KratosMultiphysics/technical-committee

File: kratos/utilities/geometry_utilities.h [hunk 3]
```diff
@@ -60,128 +55,131 @@ class GeometryUtils
...
-    //********************************************************************************
-    /**this function computes the element's volume (with sign)
+    /**
+     * @brief This function computes the element's volume (with sign)
      * @param rGeometry it is the array of nodes. It expects a triangle
+     * @deprecated This method can be replaced by geometry function without loosing performance
      */
-    static inline double CalculateVolume2D(
-        const Element::GeometryType& rGeometry)
+    KRATOS_DEPRECATED_MESSAGE("Please use the Area() method from the geometry")
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:181]: same

File: kratos/utilities/geometry_utilities.h [hunk 4]
```diff
@@ -464,369 +372,285 @@ class GeometryUtils
...
+        if(square_distance < epsilon) // near zero length line
+            return norm_2(v2); // we return the distance to the first point of line
 
-		double t = - inner_prod(v1,v2) / square_distance;
+        const double t = - inner_prod(v1,v2) / square_distance;
 
-//                KRATOS_WATCH(t);
+        if(t < 0.00) { // it is before point 1
+            // We return the distance to point 1
+            v3 = rLinePoint1 - rToPoint;
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h]: noalias

File: kratos/utilities/geometry_utilities.h [hunk 5]
```diff
@@ -464,369 +372,285 @@ class GeometryUtils
...
-		}
+            return norm_2(v3);
+        }
 
-		if(t > 1.00) // it is after point 2
-		{ // we return the distance to point 2
-			v3 = LinePoint2 - ToPoint;
+        if(t > 1.00) { // it is after point 2
+            // We return the distance to point 2
+            v3 = rLinePoint2 - rToPoint;
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h]: noalias

File: kratos/utilities/geometry_utilities.h [hunk 6]
```diff
@@ -60,128 +55,131 @@ class GeometryUtils
...
 
-    /**this function computes the element's volume (with sign)
+    /**
+     * @brief This function computes the element's volume (with sign)
      * @param rGeometry it is the array of nodes. It expects a tetrahedra
+     * @deprecated This method can be replaced by geometry function without loosing performance
      */
-    static inline double CalculateVolume3D(
-        const Element::GeometryType& rGeometry)
+    KRATOS_DEPRECATED_MESSAGE("Please use the Volume() method from the geometry")
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:114]: do the tet and triangle implement this efficiently, or do they use the generic version?

File: kratos/utilities/geometry_utilities.h [hunk 7]
```diff
@@ -250,200 +254,104 @@ class GeometryUtils
...
-		}
-		else
-		{
-			std::cout << "This is a triangle with more than two intersections!" << std::endl;
-			std::cout << "Warning: Too many intersections: " << number_of_intersection_points << std::endl;
-			std::cout << "Warning: The distances are: " << Distances << std::endl;
-
-		}
-
+        int number_of_intersection_points = CalculateTetrahedraIntersectionPoints(rGeometry, rDistances, intersection_points);
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:316]: is this large change only cosmetic?

File: kratos/utilities/geometry_utilities.h [hunk 8]
```diff
@@ -464,369 +372,285 @@ class GeometryUtils
...
      * @param TrianglePoint2 Second point of triangle
      * @param TrianglePoint3 Third point of triangle
      * @param ToPoint The point which distance is required
      * @return The distance between the point and the triangle
      */
-    static double PointDistanceToTriangle3D(Point const& TrianglePoint1,
-                                 Point const& TrianglePoint2,
-                                 Point const& TrianglePoint3,
-                                 Point const& ToPoint)
+    static double PointDistanceToTriangle3D(
```
Comment (RiccardoRossi) [inline kratos/utilities/geometry_utilities.h:479]: only cosmetic?

File: kratos/utilities/geometry_utilities.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:181]: Same as what?

File: kratos/utilities/geometry_utilities.h (hunk 8 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:479]: Yes

File: kratos/utilities/geometry_utilities.h (hunk 7 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:316]: Yes

File: kratos/utilities/geometry_utilities.h (hunk 6 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities.h:114]: The efficient one:

https://github.com/KratosMultiphysics/Kratos/blob/6b06365736c430f72eaff3ac52a0033ca39242e0/kratos/geometries/tetrahedra_3d_4.h#L404

**Review by RiccardoRossi — COMMENTED** [review]
this changes look ok to me.
@rubenzorrilla you making use of some of the functions employed here. please do the approving.

**Comment by loumalouomega** [comment]
Ping

**Comment by rubenzorrilla** [comment]
As far as I can see, this only includes cosmetic changes. Am I wrong?

**Comment by loumalouomega** [comment]
Yep

**Review by rubenzorrilla — APPROVED** [review]

**Comment by loumalouomega** [comment]
Thanks

**Pull request merged**